### PR TITLE
[Fix][Core] ray.get on a mixed list of dead and hanging tasks does not immediately raise ActorDiedError

### LINKED ISSUE #47204: [core] `ray.get` on a mixed list of dead and alive (but hanging) tasks does not immediately raise `RayActorError`
### What happened + What you expected to happen

We recently met a problem of running functions on a died actor (because the underlying node is killed). The expectation is an exception could be raised immediately. However, it seems the code just hangs. The behavior is a bit flaky, but is hangs in most cases. This problem might be a regression as the same code raise as expected in early July. 

Mini repro appended below.

### Versions / Dependencies
2.34

### Reproduction script

```
import ray
from typing import Callable

@ray.remote(num_cpus=1)
class TestClass:
    def execute(self, fn: Callable[..., None]) -> None:
        return fn()

    def exit(self):
        ray.actor.exit_actor()

def dummy_func():
    print (100)

actors = [TestClass.remote() for _ in range(10)]
for actor in actors:
    actor.exit.remote()
tasks = [actor.execute.remote(dummy_func) for actor in actors]
ray.get(tasks)
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by MortalHappiness**
## Why are these changes needed?

See the description in the corresponding issue for details.

## Related issue number

Closes: ray-project/ray#47204

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 1]
```diff
@@ -322,7 +327,7 @@ Status CoreWorkerMemoryStore::GetImpl(const std::vector<ObjectID> &object_ids,
     }
 
     // Return if all the objects are obtained.
```
Comment (rynewang) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc:329]: ...", or any existing objects are known to have exception."

**Review by rynewang — APPROVED** [review]
LGTM. Please update the comment and I will merge this. For the CI failure - please merge master after the comment update and hope it can fix the failure.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc (hunk 1 above)
Comment (MortalHappiness) [reply src/ray/core_worker/store_provider/memory_store/memory_store.cc]: done

**Comment by MortalHappiness** [comment]
@rkooo567 could you approve this PR for me? Thanks!

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**